Sources/SwiftLexicalLookup/ScopeImplementations.swift

```diff
@@ -0,0 +1,289 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+extension SyntaxProtocol {
+  /// Parent scope of this syntax node, or scope introduced by this syntax node.
+  var scope: ScopeSyntax? {
+    if let scopeSyntax = Syntax(self).asProtocol(SyntaxProtocol.self) as? ScopeSyntax {
+      scopeSyntax
+    } else {
+      self.parent?.scope
+    }
+  }
+}
+
+@_spi(Experimental) extension SourceFileSyntax: ScopeSyntax {
+  /// All names introduced in the file scope
+  /// according to the default strategy: `memberBlockUpToLastDecl`.
+  public var introducedNames: [LookupName] {
+    introducedNames(using: .memberBlockUpToLastDecl)
+  }
+
+  /// All names introduced in the file scope
+  /// using the provided configuration.
+  ///
+  /// Example usage:
+  /// ```swift
+  /// class a {}
+  /// class b {
+  ///   // <--
+  /// }
+  /// let c = 0
+  /// class d {}
+  /// if true {}
+  /// class e {}
+  /// let f = 0
+  /// ```
+  /// During lookup, according to different configurations,
+  /// names available at the marked place are:
+  /// - for `fileScopeNameIntroductionStrategy` - a, b, c, d
+  /// - for `memberBlock` - a, b, c, d, e, f
+  /// - for `codeBlock` - a
+  public func introducedNames(using fileScopeHandling: FileScopeHandlingConfig) -> [LookupName] {
+    switch fileScopeHandling {
+    case .memberBlockUpToLastDecl:
+      var encounteredNonDeclaration = false
+
+      return statements.flatMap { codeBlockItem in
+        let item = codeBlockItem.item
+
+        if encounteredNonDeclaration {
+          return LookupName.getNames(from: item, accessibleAfter: codeBlockItem.endPosition)
+        } else {
+          if item.is(DeclSyntax.self) || item.is(VariableDeclSyntax.self) {
+            return LookupName.getNames(from: item)
+          } else {
+            encounteredNonDeclaration = true
+            return LookupName.getNames(from: item, accessibleAfter: codeBlockItem.endPosition)
+          }
+        }
+      }
+    case .codeBlock:
+      return statements.flatMap { codeBlockItem in
+        LookupName.getNames(from: codeBlockItem.item, accessibleAfter: codeBlockItem.endPosition)
+      }
+    case .memberBlock:
+      return statements.flatMap { codeBlockItem in
+        LookupName.getNames(from: codeBlockItem.item)
+      }
+    }
+  }
+
+  /// Returns names matching lookup using provided file
+  /// scope handling configuration (by default: `memberBlockUpToLastDecl`).
```

Comment:
Yes, `LookupConfig` [still defaults to `.memberBlockUpToLastDecl`](https://github.com/swiftlang/swift-syntax/blob/940b06e4d8006286de1f0ba89d52b27e1368dad7/Sources/SwiftLexicalLookup/Configurations/LookupConfig.swift#L21).

Reply:
But the `config` parameter is not defaulted in the `lookup` function. So, if anything, this part of the comment should go on `LookupConfig.init` (where I don’t think it’s necessary because you can see the default value).

Reply:
Ah I get it now, that’s a good point. I mentioned what the config defaults to at `LookupConfig.init` and removed misleading comments from relevant places.